### PR TITLE
[keycloak][keycloak-gatekeeper] Initial release

### PR DESCRIPTION
**Pull request by Nuru**
## what
Initial release of [`keycloak`](https://www.keycloak.org) and `keycloak-gateeper` (formerly `keycloak-proxy`)
## why
Provide option for securing service with single-sign-on.

### CODE REVIEW

File: releases/keycloak-gatekeeper.yaml [hunk 1]
```diff
@@ -0,0 +1,113 @@
...
+#######################################################################################
+
+#
+# References:
+#   - https://github.com/gabibbo97/charts/tree/master/charts/keycloak-gatekeeper
+#   - https://hub.helm.sh/charts/gabibbo97/keycloak-gatekeeper
+#   - https://www.keycloak.org/
+#
+{{ range $index, $service := .Environment.Values.services }}
+- name: "key-gate-{{- $service.name }}"
```
Comment (osterman) [inline releases/keycloak-gatekeeper.yaml:29]: `gw` is a more common abbreviation for "gateway"
```suggestion
- name: "key-gw-{{- $service.name }}"
```

Comment (Nuru) [reply releases/keycloak-gatekeeper.yaml:29]: Yes, but this is not an abbreviation for "gateway", it is an abbreviation for "gatekeeper"

File: releases/keycloak.yaml [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+      resources:
+        limits:
+          cpu: "1000m"
+          memory: "1024Mi"
+        requests:
+          cpu: "50m"
+          memory: "512Mi"
+      persistence:
+        deployPostgres: {{ env "KEYCLOAK_DEPLOY_POSTGRES" | default "false" }}
+        dbVendor: '{{ env "KEYCLOAK_DBVENDOR" | default "h2" }}'
```
Comment (osterman) [inline releases/keycloak.yaml]:
```suggestion
        dbVendor: '{{ env "KEYCLOAK_DB_VENDOR" | default "h2" }}'
```
*nit.. camelcase to snakecase

File: releases/keycloak-gatekeeper.yaml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+        {{- if index $service "portalName"  }}
+        forecastle.stakater.com/expose: "true"
+        forecastle.stakater.com/appName: "{{ $service.portalName }}"
+        {{- end }}
+      hosts:
+      - {{ $service.host }}
+      # tls:
+      #  - secretName: keycloak-http-tls
+      #    hosts:  [{{ env "KEYCLOAK_INGRESS_HOSTS" }}]
+    discoveryURL: '{{- requiredEnv "KOPS_OIDC_ISSUER_URL" }}'
```
Comment (osterman) [inline releases/keycloak-gatekeeper.yaml:68]: This only applies to the kubernetes dashboard, no?

Comment (Nuru) [reply releases/keycloak-gatekeeper.yaml:68]: No, every instance of Gatekeeper needs to communicate with Keycloak

File: releases/keycloak-gatekeeper.yaml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+    upstreamURL: '{{- $service.upstream }}'
+    skipUpstreamTlsVerify: {{ index $service "skipUpstreamTlsVerify" | default "false" }}
+    ClientID: 'kubernetes'
+    ClientSecret: '{{- requiredEnv "KEYCLOAK_GATEKEEPER_DASHBOARD_CLIENT_SECRET" }}'
```
Comment (osterman) [inline releases/keycloak-gatekeeper.yaml]: Not just `KEYCLOAK_GATEKEEPER_CLIENT_SECRET `?

File: releases/forecastle.yaml [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+          image:
+            pullPolicy: "IfNotPresent"
+          namespace: "kube-system"
+          deployment:
+            annotations: ""
+          config:
+            namespaces: [{{- env "FORECASTLE_NAMESPACES" | default "kube-system,monitoring" }}]
+            title: {{ env "FORECASTLE_TITLE" | default "Portal" }}
+          service:
+            annotations: ""
```
Comment (osterman) [inline releases/forecastle.yaml:43]: is this correct? usually annotations is a list `[]`

File: releases/keycloak-gatekeeper.yaml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+#      host: forecastle.us-west-2.corp.cpco.io
+#      upstream: http://forecastle.kube-system.svc.cluster.local
+#      rules:
+#        - "uri=/*|roles=kube-admin,user|require-any-role=true"
+######
+
+#
+environments:
+  default:
+    values:
```
Comment (osterman) [inline releases/keycloak-gatekeeper.yaml:12]: We've been sticking values here: 

https://github.com/cloudposse/helmfiles/tree/master/releases/values

Comment (Nuru) [reply releases/keycloak-gatekeeper.yaml:12]: These values go in Geodesic, they are not part of the release

File: releases/forecastle.yaml (hunk 5 above)
Comment (Nuru) [reply releases/forecastle.yaml:43]: This is correct. Annotations are usually a map/dict. The chart has a lot of default annotations and an empty map did not suppress them.

File: releases/keycloak-gatekeeper.yaml (hunk 6 above)
Comment (osterman) [reply releases/keycloak-gatekeeper.yaml:12]: right, but the user should stick them in `/conf/helmfiles/releases/values`

Comment (osterman) [reply releases/keycloak-gatekeeper.yaml:12]: if you look in `values/` those are all *examples* designed to be overridden

Comment (Nuru) [reply releases/keycloak-gatekeeper.yaml:12]: Well, at this point `values/` is mostly templates that are used by the helmfiles, but I can move this if you want.

File: releases/keycloak-gatekeeper.yaml [hunk 7]
```diff
@@ -0,0 +1,95 @@
...
+        kubernetes.io/tls-acme: "{{ index $service "useTLS" | default "false" }}"
+        external-dns.alpha.kubernetes.io/target: '{{ requiredEnv "NGINX_INGRESS_HOSTNAME" }}'
+        external-dns.alpha.kubernetes.io/ttl: "60"
+        {{- if index $service "portalName"  }}
+        forecastle.stakater.com/expose: "true"
+        forecastle.stakater.com/appName: "{{ $service.portalName }}"
+        {{- end }}
+      hosts:
+      - {{ $service.host }}
+      # tls:
```
Comment (osterman) [inline releases/keycloak-gatekeeper.yaml]: this cannot be overriden, so should we keep the comment?

**Review by osterman — APPROVED** [review]

**Pull request merged**